keps/sig-api-machinery/2876-crd-validation-expression-language/README.md

```diff
@@ -348,25 +348,106 @@ to fail with a descriptive error.
+- Regex: `matches`
+- Date/time accessors: `getDate`, `getDayOfMonth`, `getDayOfWeek`, `getDayOfYear`, `getFullYear`, `getHours`, `getMilliseconds`, `getMinutes`, `getMonth`, `getSeconds`
+
+CEL standard [macros](https://github.com/google/cel-spec/blob/master/doc/langdef.md#macros) include:
+
+- `has`
+- `all`
+- `exists`
+- `exists_one`
+- `map`
+- `filter`
+
+CEL [extended string function library](https://github.com/google/cel-go/blob/master/ext/strings.go) includes:
+- 
+- `charAt`
+- `indexOf`, `lastIndexOf`
+- `upperAscii`, `lowerAscii`
+- `replace` (string literal replacement, not regex)
+- `split`
+- `substring`
+- `trim`
+
+Notable absences from the standard functions, when comparing against the core libraries of Go, Java and Python are:
+
+Strings:
+
+- trim/trimLeft/trimRight (trimming either by whitespace or by provided cutset)
+- trimPrefix/trimSuffix
+- join
+- replace (regex)
+- format (e.g. `fmt.Sprintf()`)
+
+Lists:
+
+- sort
+- indexOf/lastIndexOf
+- replace
+- sublist
+- split
+- reverse
+- reduce or sum/min/max (reduce is more general, the others are more concrete)
+
+Numbers:
+
+- abs/cel/floor/round
+- exp/log/log10/logN/pow/sqrt
+- max(x, y, ..)/min(x, y, ...)
+- <trig functions>
+
+Maps:
+
+- <Ability to construct a map using a comprehension>
+
+Some considerations when selecting which of the above we should include in CEL expressions in Kubernetes:
+
+- We should include functions that will be needed for future use cases of CEL in Kubernetes (e.g. general admission) as
+  ones needed for validation.
+- Regex replace is very powerful and useful. It is also potentially dangerous due to its ability to allocate memory.
+- `format`: Since CEL supports string concatenation, the value of having format would only be to do things like format floats.
+  Formatting functions are complex and require extensive documentation to teach.
+- `sort` makes it possible to check if a list is sorted. It can be an expensive operation.
+- `indexOf` / `lastIndexOf` / `split` / `replace` (the string functions) could be overloaded to provide the equivalent list functions?
+   Keeping string and list functions consistent seems like a good way to keep the learning curve down.
+- `reduce` is going to non-obvious to anyone that doesn't have a functional programming background?
+  - CEL already provides `map`..
+  - I don't want to ever be expected to expose `fold` (or `zip` either, I think)
+- "Ability to construct a map using a comprehension": this appears mechanically problematic to support in CEL?
+
+Proposal:
+
+- Keep all the CEL standard functions and macros as well as the extended string library.
+- Add regex replace support. It's very useful, and we already allow regex matching.
+- Add `trim` / `trimLeft` / `trimRight` (overloaded to take an optional cutset arg) for strings
+- Add `trimPrefix` / `trimSuffix` for strings
+- Add `sort` for lists with comparable elements
+- Add a `reduce` macro? (I'm on the fence on this one)
```

Comment:
What potential usecase do you see for this?

Reply:
Honestly, with the plan to provide `sum`, `max` and `min` functions directly, I can't immediately think of any use cases of reduce. I'll drop it from this list now.